packages/emf/backend/sirius-components-interpreter/src/main/java/org/eclipse/sirius/components/interpreter/AQLInterpreter.java

```diff
@@ -112,12 +114,38 @@ public AQLInterpreter(List<Class<?>> classes, List<Object> instances, List<EPack
         ePackages.stream().filter(this::isValidEPackage).forEach(ePackage -> {
             String nsURI = ePackage.getNsURI();
             long start = System.nanoTime();
-            this.queryEnvironment.registerEPackage(ePackage);
+            AtomicBoolean cacheHit = new AtomicBoolean();
+
+            EPackage registeredEPackage = ((EPackageProvider) this.queryEnvironment.getEPackageProvider()).registerPackage(ePackage);
+            if (registeredEPackage != null) {
+                var optionalCachedServices = ePackage.eAdapters().stream()
```

Comment:
Isn't that a bit brittle to attach an adapter like that to the ePackage? We can move forward like that for now but I'm not sure that keeping this state of services in the EPackage like that is threadsafe. It may be easier to compute that when we load the server (just like we are registering the static EPackages).